File: adgeistkit/src/main/java/com/adgeistkit/ads/AdActivity.java
package com.adgeistkit.ads;

import android.graphics.Rect;
import android.os.Handler;
import android.os.Looper;
import android.os.SystemClock;
import android.util.Log;
import android.view.View;
import android.view.ViewTreeObserver;
import android.webkit.WebView;
import android.widget.HorizontalScrollView;
import android.widget.ScrollView;

import androidx.annotation.Nullable;
import androidx.core.widget.NestedScrollView;

public class AdActivity {
    private static final String TAG = "Ad Activity";
    private static final double VISIBILITY_THRESHOLD = 0.5f;
    private static final long MIN_VIEW_TIME = 1000L;

    private final BaseAdView baseAdView;
    private final long renderStartTime;

    private final String mediaType;
    private float currentVisibilityRatio;
    private boolean isVisible;
    private long viewStartTime;
    private long totalViewTime;
    private boolean hasViewEvent = false;
    private boolean hasImpression = false;
    private long renderTime;

    private long playbackStartTime = 0;
    private long totalPlaybackTime = 0;
    private boolean hasEnded = false;
    private boolean hasSentPlaybackEvent = false;

    private ViewTreeObserver.OnScrollChangedListener scrollListener;
    private ViewTreeObserver.OnWindowFocusChangeListener focusListener;

    private final Handler handler = new Handler(Looper.getMainLooper());
    private Runnable visibilityCheckRunnable;

    public AdActivity(BaseAdView baseAdView){
        this.baseAdView = baseAdView;
        this.renderStartTime = SystemClock.elapsedRealtime();
        this.mediaType = baseAdView.mediaType;
        initialize();
    }

    private void initialize(){
        setupVisibilityTracking();
    }

    private void setupVisibilityTracking(){
        ViewTreeObserver vto = baseAdView.getViewTreeObserver();

        scrollListener = this::checkVisibility;
        vto.addOnScrollChangedListener(scrollListener);

        focusListener = hasFocus -> onVisibilityChange(hasFocus);
        vto.addOnWindowFocusChangeListener(focusListener);

        checkVisibility();
    }

    private void checkVisibility() {
        Rect rect = new Rect();
        boolean isVisible = baseAdView.getGlobalVisibleRect(rect);

        if (!isVisible) {
            handleVisibilityChange(false);
            return;
        }

        int totalWidth = baseAdView.getWidth();
        int totalHeight = baseAdView.getHeight();

        if (totalWidth == 0 || totalHeight == 0) {
            return;
        }

        int visibleWidth = rect.width();
        int visibleHeight = rect.height();

        currentVisibilityRatio = (visibleWidth * visibleHeight) / (float) (totalWidth * totalHeight);
        boolean newVisible = currentVisibilityRatio >= VISIBILITY_THRESHOLD;

        handleVisibilityChange(newVisible);
    }

    private void handleVisibilityChange(boolean newVisible) {
        boolean wasVisible = isVisible;
        isVisible = newVisible;

        if (isVisible && !wasVisible) {
            if (viewStartTime == 0) {
                viewStartTime = SystemClock.elapsedRealtime();
                startVisibilityCheck();
            }
            if ("video".equals(mediaType) && !hasEnded) {
                getWebView().onResume();
                onVideoPlay();
            }
        } else if (!isVisible && wasVisible) {
            updateViewTime();
            stopVisibilityCheck();
            if ("video".equals(mediaType) && !hasEnded) {
                getWebView().onPause();
                onVideoPause();
            }
        }
    }

    private void startVisibilityCheck() {
        if (visibilityCheckRunnable != null) return;
        visibilityCheckRunnable = new Runnable() {
            @Override
            public void run() {
                if (isVisible && viewStartTime > 0 && !hasViewEvent) {
                    long timeInView = SystemClock.elapsedRealtime() - viewStartTime;
                    if (timeInView >= MIN_VIEW_TIME) {
                        hasViewEvent = true;
                        baseAdView.listener.onAdImpression();
                        float scrollDepth = getScrollDepth();
                        long timeToVisible = SystemClock.elapsedRealtime() - renderStartTime;
                        Log.e(TAG, " scrollDepth=" + scrollDepth + "timeToVisible =" +   timeToVisible + "currentVisiblityRatio =" + currentVisibilityRatio);
                        stopVisibilityCheck();
                    }
                }
                handler.postDelayed(this, 100);
            }
        };
        handler.post(visibilityCheckRunnable);
    }

    private View findRootScrollView(View view) {
        View scrollParent = null;

        while (view != null) {
            if (isScrollableView(view)) {
                scrollParent = view;
            }
            if (view.getParent() instanceof View) {
                view = (View) view.getParent();
            } else {
                break;
            }
        }
        return scrollParent;
    }

    private boolean isScrollableView(View view) {
        return (view instanceof ScrollView
                || view instanceof NestedScrollView
                || view instanceof HorizontalScrollView);
    }

    public float getScrollDepth() {
        View scrollView = findRootScrollView(baseAdView);

        // Get AdView absolute Y on screen
        int[] adLocation = new int[2];
        baseAdView.getLocationOnScreen(adLocation);
        int adTopOnScreen = adLocation[1];

        // Get ScrollView top on screen
        int[] scrollLocation = new int[2];
        scrollView.getLocationOnScreen(scrollLocation);
        int scrollTopOnScreen = scrollLocation[1];

        int requiredScroll = adTopOnScreen - scrollTopOnScreen;
        if (requiredScroll <= 0) {
            return 1f;
        }

        // 4. Current scroll amount
        int currentScroll = scrollView.getScrollY();
        float rawRatio = (float) currentScroll / (float) requiredScroll;
        if (rawRatio < 0f) rawRatio = 0f;
        if (rawRatio > 1f) rawRatio = 1f;

        return rawRatio;
    }

    private void updateViewTime() {
        if (viewStartTime > 0) {
            totalViewTime += SystemClock.elapsedRealtime() - viewStartTime;
            viewStartTime = 0;
            stopVisibilityCheck();
        }
    }

    private void stopVisibilityCheck() {
        if (visibilityCheckRunnable != null) {
            handler.removeCallbacks(visibilityCheckRunnable);
            visibilityCheckRunnable = null;
        }
    }

    public void onVisibilityChange(boolean hasFocus) {
        if (!hasFocus) {
            updateViewTime();
            stopVisibilityCheck();
            if ("video".equals(mediaType) && !hasEnded) {
                getWebView().onPause();
                onVideoPause();
            }
        } else if (isVisible) {
            if ("video".equals(mediaType) && !hasEnded) {
                getWebView().onResume();
                onVideoPlay();
            }
        }
    }

    public void onVideoPlay() {
        if (playbackStartTime == 0) {
            playbackStartTime = SystemClock.elapsedRealtime();
        }
    }

    public void onVideoPause() {
        updatePlaybackTime();
    }

    public void onVideoEnd() {
        if (!hasEnded && "video".equals(mediaType)) {
            hasEnded = true;
            updatePlaybackTime();
        }
    }

    private void updatePlaybackTime() {
        if (playbackStartTime > 0 && "video".equals(mediaType)) {
            totalPlaybackTime += SystemClock.elapsedRealtime() - playbackStartTime;
            playbackStartTime = 0;
        }
    }

    public void captureImpression() {
        if (!hasImpression) {
            renderTime = SystemClock.elapsedRealtime() - renderStartTime;
            Log.e(TAG, "renderTime= " + renderTime);
            baseAdView.listener.onAdLoaded();
            hasImpression = true;
        }
    }

    public void captureClick(){
        baseAdView.listener.onAdClicked();
    }

    public void captureTotalViewTime(){

    }

    public void captureTotalVideoPlaybackTime(){
        if (totalPlaybackTime > 0 && !hasSentPlaybackEvent && "video".equals(mediaType)) {
            hasSentPlaybackEvent = true;
        }
    }

    @Nullable
    private WebView getWebView(){
        if (baseAdView.getChildCount() > 0 && baseAdView.getChildAt(0) instanceof WebView) {
            return (WebView) baseAdView.getChildAt(0);
        }
        return null;
    }

    public void destroy(){
        ViewTreeObserver vto = baseAdView.getViewTreeObserver();
        if(vto.isAlive()){
            vto.removeOnScrollChangedListener(scrollListener);
        }

        captureTotalViewTime();
        captureTotalVideoPlaybackTime();
        updateViewTime();
        stopVisibilityCheck();
    }
}

File: adgeistkit/src/main/java/com/adgeistkit/ads/AdActivity.kt
package com.adgeistkit.ads

import android.graphics.Rect
import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.util.Log
import android.view.View
import android.view.ViewTreeObserver.OnScrollChangedListener
import android.view.ViewTreeObserver.OnWindowFocusChangeListener
import android.webkit.WebView
import android.widget.HorizontalScrollView
import android.widget.ScrollView
import androidx.core.widget.NestedScrollView
import com.adgeistkit.AdgeistCore.Companion.getInstance
import com.adgeistkit.ads.network.AnalyticsRequest

class AdActivity(private val baseAdView: BaseAdView) {
    private val postCreativeAnalytics = getInstance().postCreativeAnalytics()

    private val renderStartTime = SystemClock.elapsedRealtime()

    private val mediaType = baseAdView.mediaType
    private var currentVisibilityRatio = 0f
    private var isVisible = false
    private var viewStartTime: Long = 0
    private var totalViewTime: Long = 0
    private var hasViewEvent = false
    private var hasImpression = false
    private var renderTime: Long = 0

    private var playbackStartTime: Long = 0
    private var totalPlaybackTime: Long = 0
    private var hasEnded = false
    private var hasSentPlaybackEvent = false

    private var scrollListener: OnScrollChangedListener? = null
    private var focusListener: OnWindowFocusChangeListener? = null

    private val handler = Handler(Looper.getMainLooper())
    private var visibilityCheckRunnable: Runnable? = null

    init {
        initialize()
    }

    private fun initialize() {
        setupVisibilityTracking()
    }

    private fun setupVisibilityTracking() {
        val vto = baseAdView.viewTreeObserver

        scrollListener = OnScrollChangedListener { this.checkVisibility() }
        vto.addOnScrollChangedListener(scrollListener)

        focusListener =
            OnWindowFocusChangeListener { hasFocus: Boolean -> onVisibilityChange(hasFocus) }
        vto.addOnWindowFocusChangeListener(focusListener)

        checkVisibility()
    }

    private fun checkVisibility() {
        val rect = Rect()
        val isVisible = baseAdView.getGlobalVisibleRect(rect)

        if (!isVisible) {
            handleVisibilityChange(false)
            return
        }

        val totalWidth = baseAdView.width
        val totalHeight = baseAdView.height

        if (totalWidth == 0 || totalHeight == 0) {
            return
        }

        val visibleWidth = rect.width()
        val visibleHeight = rect.height()

        currentVisibilityRatio =
            (visibleWidth * visibleHeight) / (totalWidth * totalHeight).toFloat()
        val newVisible = currentVisibilityRatio >= VISIBILITY_THRESHOLD

        handleVisibilityChange(newVisible)
    }

    private fun handleVisibilityChange(newVisible: Boolean) {
        val wasVisible = isVisible
        isVisible = newVisible

        if (isVisible && !wasVisible) {
            if (viewStartTime == 0L) {
                viewStartTime = SystemClock.elapsedRealtime()
                startVisibilityCheck()
            }
            if ("video" == mediaType && !hasEnded) {
                webView?.onResume()
                onVideoPlay()
            }
        } else if (!isVisible && wasVisible) {
            updateViewTime()
            stopVisibilityCheck()
            if ("video" == mediaType && !hasEnded) {
                webView?.onPause()
                onVideoPause()
            }
        }
    }

    private fun startVisibilityCheck() {
        if (visibilityCheckRunnable != null) return
        visibilityCheckRunnable = object : Runnable {
            override fun run() {
                if (isVisible && viewStartTime > 0 && !hasViewEvent) {
                    val timeInView = SystemClock.elapsedRealtime() - viewStartTime
                    if (timeInView >= MIN_VIEW_TIME) {
                        hasViewEvent = true
                        baseAdView.listener?.onAdImpression()
                        val scrollDepth: Float = scrollDepth()
                        val timeToVisible = SystemClock.elapsedRealtime() - renderStartTime
                        val analyticsRequest: AnalyticsRequest =
                            AnalyticsRequest.AnalyticsRequestBuilder(baseAdView.metaData, baseAdView.isTestMode)
                                .trackViewableImpression(
                                    timeToVisible,
                                    scrollDepth,
                                    currentVisibilityRatio,
                                    timeInView
                                )
                                .build()
                        postCreativeAnalytics.sendTrackingDataV2(analyticsRequest)
                        stopVisibilityCheck()
                    }
                }
                handler.postDelayed(this, 100)
            }
        }

        val runnable = visibilityCheckRunnable
        handler.post(runnable!!)
    }

    private fun findRootScrollView(view: View?): View? {
        var view = view
        var scrollParent: View? = null

        while (view != null) {
            if (isScrollableView(view)) {
                scrollParent = view
            }
            if (view.parent is View) {
                view = view.parent as View
            } else {
                break
            }
        }
        return scrollParent
    }

    private fun isScrollableView(view: View): Boolean {
        return (view is ScrollView
                || view is NestedScrollView
                || view is HorizontalScrollView)
    }

    private fun scrollDepth(): Float {
        val scrollView = findRootScrollView(baseAdView)

        // Get AdView absolute Y on screen
        val adLocation = IntArray(2)
        baseAdView.getLocationOnScreen(adLocation)
        val adTopOnScreen = adLocation[1]

        // Get ScrollView top on screen
        val scrollLocation = IntArray(2)
        scrollView!!.getLocationOnScreen(scrollLocation)
        val scrollTopOnScreen = scrollLocation[1]

        val requiredScroll = adTopOnScreen - scrollTopOnScreen
        if (requiredScroll <= 0) {
            return 1f
        }

        // 4. Current scroll amount
        val currentScroll = scrollView.scrollY
        var rawRatio = currentScroll.toFloat() / requiredScroll.toFloat()
        if (rawRatio < 0f) rawRatio = 0f
        if (rawRatio > 1f) rawRatio = 1f

        return rawRatio
    }

    private fun updateViewTime() {
        if (viewStartTime > 0) {
            totalViewTime += SystemClock.elapsedRealtime() - viewStartTime
            viewStartTime = 0
            stopVisibilityCheck()
        }
    }

    private fun stopVisibilityCheck() {
        if (visibilityCheckRunnable != null) {
            handler.removeCallbacks(visibilityCheckRunnable!!)
            visibilityCheckRunnable = null
        }
    }

    fun onVisibilityChange(hasFocus: Boolean) {
        if (!hasFocus) {
            updateViewTime()
            stopVisibilityCheck()
            if ("video" == mediaType && !hasEnded) {
                webView?.onPause()
                onVideoPause()
            }
        } else if (isVisible) {
            if ("video" == mediaType && !hasEnded) {
                webView?.onResume()
                onVideoPlay()
            }
        }
    }

    fun onVideoPlay() {
        if (playbackStartTime == 0L) {
            playbackStartTime = SystemClock.elapsedRealtime()
        }
    }

    fun onVideoPause() {
        updatePlaybackTime()
    }

    fun onVideoEnd() {
        if (!hasEnded && "video" == mediaType) {
            hasEnded = true
            updatePlaybackTime()
        }
    }

    private fun updatePlaybackTime() {
        if (playbackStartTime > 0 && "video" == mediaType) {
            totalPlaybackTime += SystemClock.elapsedRealtime() - playbackStartTime
            playbackStartTime = 0
        }
    }

    fun captureImpression() {
        if (!hasImpression) {
            renderTime = SystemClock.elapsedRealtime() - renderStartTime
            Log.e(
                TAG,
                "renderTime= $renderTime"
            )
            baseAdView.listener?.onAdLoaded()
            val analyticsRequest: AnalyticsRequest =
                AnalyticsRequest.AnalyticsRequestBuilder(baseAdView.metaData, baseAdView.isTestMode)
                    .trackImpression(renderTime)
                    .build()
            postCreativeAnalytics.sendTrackingDataV2(analyticsRequest)
            hasImpression = true
        }
    }

    fun captureClick() {
        baseAdView.listener?.onAdClicked()
        val analyticsRequest: AnalyticsRequest =
            AnalyticsRequest.AnalyticsRequestBuilder(baseAdView.metaData, baseAdView.isTestMode)
                .trackClick()
                .build()
        postCreativeAnalytics.sendTrackingDataV2(analyticsRequest)
    }

    fun captureTotalViewTime() {
        val analyticsRequest: AnalyticsRequest =
            AnalyticsRequest.AnalyticsRequestBuilder(baseAdView.metaData, baseAdView.isTestMode)
                .trackTotalViewTime(totalViewTime)
                .build()
        postCreativeAnalytics.sendTrackingDataV2(analyticsRequest)
    }

    fun captureTotalVideoPlaybackTime() {
        if (totalPlaybackTime > 0 && !hasSentPlaybackEvent && "video" == mediaType) {
            val analyticsRequest: AnalyticsRequest =
                AnalyticsRequest.AnalyticsRequestBuilder(baseAdView.metaData, baseAdView.isTestMode)
                    .trackTotalPlaybackTime(totalPlaybackTime)
                    .build()
            postCreativeAnalytics.sendTrackingDataV2(analyticsRequest)
            hasSentPlaybackEvent = true
        }
    }

    private val webView: WebView?
        get() {
            if (baseAdView.childCount > 0 && baseAdView.getChildAt(0) is WebView) {
                return baseAdView.getChildAt(0) as WebView
            }
            return null
        }

    fun destroy() {
        val vto = baseAdView.viewTreeObserver
        if (vto.isAlive) {
            vto.removeOnScrollChangedListener(scrollListener)
        }

        captureTotalViewTime()
        captureTotalVideoPlaybackTime()
        updateViewTime()
        stopVisibilityCheck()
    }

    companion object {
        private const val TAG = "Ad Activity"
        private const val VISIBILITY_THRESHOLD = 0.5
        private const val MIN_VIEW_TIME = 1000L
    }
}